src/qcrypto/simbasics.py

```diff
@@ -1,132 +1,424 @@
+        outcome = np.random.choice([0, 1], p=[probs_0, probs_1])
+        self._update_state_post_measurement(qubit_idx, outcome)
+        return outcome
 
-    def project(self, qubit, base):
+    def measure_all(self, order):
         """
-        Projects the given qubit unto the agent's measuring basis
+        Measures all of the qubits
+
+        Args:
+            order (str): Specifies the order in which the qubits will be measured.
+                "simult" = all qubits measured simultaneously
+                "sequential" = qubits measured in sequential order (first 0, second 1, etc.)
+
+        Returns:
+            Outcome of the measurements done. Array of 0's and 1's equal in length to the number of qubits in the system.
         """
-        newqubit = qubit
-        base_delta = base - newqubit.base
+        if order == "simult":
+            outcome = np.random.choice(
+                np.arange(len(self._state)), p=self._calculate_measurement_probs()
+            )
+            self._state.fill(0 + 0j)
+            self._state[outcome] = 1 + 0j
+            outcome_arr = np.array(
+                list(
+                    "0" * (self.num_qubits - len(bin(outcome)[2:])) + bin(outcome)[2:]
+                ),
+                dtype=int,
+            )
+            return outcome_arr
+        elif order == "sequential":
+            outcome = []
+            for i in range(self.num_qubits):
+                outcome.append(self.measure(qubit_idx=i))
+            self._update_state_post_measurement
+            outcome_arr = np.array(outcome)
+            return outcome_arr
+        else:
+            raise ValueError("Order specified not valid.")
 
-        # Projection oprator
-        proj_mtrx = np.array(
-            [
-                [np.cos(base_delta), -1 * np.sin(base_delta)],
-                [np.sin(base_delta), np.cos(base_delta)],
-            ]
-        )
+    def _calculate_measurement_probs(self, qubit_idx: int = None):
+        """
+        From the probability amplitude, computes the probability that a measurement of a given qubit will give 0 or 1.
 
-        newqubit.state = np.dot(proj_mtrx, newqubit.state)
-        newqubit.normalize()
+        Args:
+            qubits_idx (int): Index identifying the qubit to be measured
 
-        newqubit.base = base
+        Returns:
+            Probability of measuring qubit in position qubit_idx to be measured to be 0 or to be 1
+        """
+        if qubit_idx is None:
+            outcome_probs = np.abs(self._state) ** 2
+            return outcome_probs
+        else:
+            if qubit_idx >= self.num_qubits or qubit_idx < 0:
+                raise ValueError(
+                    "Invalid qubit index. Make sure it is between 1 and num_qubits - 1"
+                )
 
-        return newqubit
+            prob_0 = 0
+            prob_1 = 0
+            for idx, prob_amp in enumerate(self._state):
+                if (idx >> qubit_idx) & 1 == 0:
```

Comment:
Never seen `>>` as an operator before. Is this the [shift operator](https://www.prepbytes.com/blog/python/right-shift-operator-in-python/#:~:text=In%20Python%2C%20the%20right%20shift,want%20to%20shift%20the%20number.) (a bitwise operator)? Was it supposed to be `>`?

Reply:
Yep, its the bitwise shift operator.